src/highlighter/highlighter.ts

```diff
@@ -0,0 +1,59 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Adapter for using server-side highlighter code.
+
+import { NodeFront } from "protocol/thread/node";
+import { NodeBoundsFront } from "protocol/thread/bounds";
+import { BoxModelHighlighter } from "devtools/server/actors/highlighters/box-model";
+
+class Highlighter {
+  private boxModelHighlighter: BoxModelHighlighter | undefined;
+  currentNode: NodeFront | NodeBoundsFront | null = null;
+
+  async highlight(node: NodeFront | NodeBoundsFront) {
+    if (!node) {
+      return;
+    }
+    this.currentNode = node;
+    if ("getBoxModel" in node) {
+      await node.getBoxModel();
+    }
+    if (this.currentNode !== node) {
+      return;
+    }
+
+    this.maybeDestroyBrokenHighlighter();
+    this.maybeCreateNewHighlighter();
+
+    this.boxModelHighlighter?.show(node);
+  }
+
+  unhighlight() {
+    this.maybeDestroyBrokenHighlighter();
+    this.boxModelHighlighter?.hide();
+    this.currentNode = null;
+  }
+
+  private maybeDestroyBrokenHighlighter() {
```

Comment:
This looks like it's wrapping some tricky edge-cases that you have run into. Personally, if I were to run into this code in the codebase without tests or a PR description about why it's here and how it works, I would be very scared to touch it for fear that I don't know the particular edge cases it's addressing. It seems like as a team Replay isn't super interested in documenting things via commit messages, but I wonder if there is a place where we could put this kind of knowledge/documentation?

Reply:
Also, sorry if that is a stupid question, I don't know anything about how the node picker works 😅